### PR TITLE
feat: add BBS+ support to crypto.Crypto and webkms

### LINKED ISSUE #2295: Support BBS+ in KMS
@Baha-sk :
>We can create a new Tink key for BBS+, but signing should happen through the key’s primitive, ie signature logic should be internal to the key package

>One thing we need to investigate is to check  is weather to update Tink Signature key type with new key manager (like ecdsa or ed25519 signature managers in Tink) or invent a fully new Tink key package in our framework for secp256k1 and BBS+

### PR DESCRIPTION
**Pull request by Baha-sk**
closes #2295

Signed-off-by: Baha Shaaban <baha.shaaban@securekey.com>

### CODE REVIEW

**Review by fqutishat — APPROVED** [review]

File: pkg/crypto/api.go [hunk 1]
```diff
@@ -62,6 +60,28 @@ type Crypto interface {
 	// 		unwrapped key in raw bytes
 	// 		error in case of errors
 	UnwrapKey(recWK *RecipientWrappedKey, kh interface{}, opts ...WrapKeyOpts) ([]byte, error)
+	// BBSSign will sign create signature of each message in messages and aggregate it into a single signature using a
+	// matching signature primitive in kh key handle of a private key.
+	// returns:
+	// 		signature in []byte
+	//		error in case of errors
+	BBSSign(messages [][]byte, kh interface{}) ([]byte, error)
```
Comment (troyronda) [inline pkg/crypto/api.go:68]: It’s weird to call this BBS - make a generic name in an interface.

File: pkg/crypto/api.go [hunk 2]
```diff
@@ -62,6 +60,28 @@ type Crypto interface {
...
+	// BBSVerify will verify an aggregated signature of one or more messages using a matching signature primitive in kh
+	//  key handle of a public key
+	// returns:
+	// 		error in case of errors or nil if signature verification was successful
+	BBSVerify(messages [][]byte, signature []byte, kh interface{}) error
```
Comment (troyronda) [inline pkg/crypto/api.go:73]: Same here.

File: pkg/crypto/api.go (hunk 1 above)
Comment (Baha-sk) [reply pkg/crypto/api.go:68]: Sign() is already used

File: pkg/crypto/api.go (hunk 2 above)
Comment (Baha-sk) [reply pkg/crypto/api.go:73]: Verify() is already used

File: pkg/crypto/api.go (hunk 1 above)
Comment (Baha-sk) [reply pkg/crypto/api.go]: updated to SignMulti()

File: pkg/crypto/api.go (hunk 2 above)
Comment (Baha-sk) [reply pkg/crypto/api.go]: Updated to VerifyMulti()

File: pkg/crypto/webkms/remotecrypto.go [hunk 3]
```diff
@@ -103,6 +129,12 @@ const (
 	verifyMACURI  = "/verifymac"
 	wrapURI       = "/wrap"
 	unwrapURI     = "/unwrap"
+
+	// BBS+ endpoints.
```
Comment (troyronda) [inline pkg/crypto/webkms/remotecrypto.go:133]: These should be generic rather than BBS specific.

File: pkg/crypto/webkms/remotecrypto.go [hunk 4]
```diff
@@ -58,15 +58,41 @@ type signReq struct {
...
+	Signature       string   `json:"signature,omitempty"`
+	Nonce           string   `json:"nonce,omitempty"`
+	RevealedIndexes []int    `json:"revealedIndexes,omitempty"`
+}
+
 type signResp struct {
 	Signature string `json:"signature,omitempty"`
 }
 
+type bbsDeriveProofResp struct {
```
Comment (troyronda) [inline pkg/crypto/webkms/remotecrypto.go:76]: Keep these generically named.

File: pkg/crypto/webkms/remotecrypto.go [hunk 5]
```diff
@@ -586,6 +618,195 @@ func (r *RemoteCrypto) UnwrapKey(recWK *crypto.RecipientWrappedKey, keyURL inter
...
+	if err != nil {
+		return nil, fmt.Errorf("unmarshal signature for BBS+ Sign failed [%s, %w]", destination, err)
+	}
+
+	keyBytes, err := base64.URLEncoding.DecodeString(httpResp.Signature)
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Infof("overall BBS+ Sign duration: %s", time.Since(startSign))
```
Comment (troyronda) [inline pkg/crypto/webkms/remotecrypto.go:668]: Todo: these shouldn't be info logs.

File: pkg/crypto/webkms/remotecrypto.go (hunk 3 above)
Comment (Baha-sk) [reply pkg/crypto/webkms/remotecrypto.go]: updated

File: pkg/crypto/webkms/remotecrypto.go (hunk 4 above)
Comment (Baha-sk) [reply pkg/crypto/webkms/remotecrypto.go]: renamed

File: pkg/crypto/webkms/remotecrypto.go (hunk 5 above)
Comment (Baha-sk) [reply pkg/crypto/webkms/remotecrypto.go:678]: we'll switch to Debugf once perf testing is complete in a future commit

**Review by troyronda — APPROVED** [review]

**Pull request merged**